src/poliastro/twobody/events.py

```diff
@@ -25,8 +30,84 @@ def terminal(self):
     def last_t(self):
         return self._last_t * u.s
 
-    def __call__(self, t, u, k):
+    def __call__(self, t, u_, k):
         self._last_t = t
-        H = norm(u[:3])
+        H = norm(u_[:3])
         # SciPy will search for H - R = 0
         return H - self._R
+
+
+class PenumbraEvent:
+    """Detect whether a satellite is in penumbra or not.
+
+    Parameters
+    ----------
+    r_sun: ~astropy.units.Quantity
+        Position vector of the Sun with respect to the Earth.
+    R: ~astropy.units.Quantity
+        Radius of the attractor.
+
+    """
+
+    def __init__(self, r_sun, r_s, r_p, terminal=False):
+        self._r_sun = r_sun
+        self._r_s = r_s
+        self._r_p = r_p
+        self._last_t = None
+        self._terminal = terminal
+
+    @property
+    def terminal(self):
+        return self._terminal
+
+    @property
+    def last_t(self):
+        return self._last_t * u.s
+
+    def __call__(self, t, u_, k):
+        self._last_t = t
+        r_sat = (u_[:3] << u.km).value
+        r_sun = self._r_sun.to(u.km).value
+        r_s = self._r_s.to(u.km).value
+        r_p = self._r_p.to(u.km).value
+        is_in_penumbra = is_in_penumbral_shadow_fast(r_sat, r_sun, r_s, r_p)
+
+        return 0 if is_in_penumbra == 0 else 1
```

Comment:
I did this for the `return` condition inside `__call__` since `is_in_penumbra` returns the difference between vertical distances if it is in one of the shadow regions (umbra or penumbra) and returns the satellite's position vector if it is not in one of those shadow regions.

The reason for doing this was `r_sat_norm`, I thought, would never be zero since it is the distance from the center of the primary attractor and satellite would never go to the center of the Earth, hopefully :)

Importantly, doing, for example:

```py
return is_in_penumbra
```
didn't seem to work (the same bug: satellite's position vector hangs) and returning True or False inside `is_in_penumbral_shadow` also didn't seem to work for some reason.

Reply:
I understand the idea behind returning the `r_sat_norm` but do not agree with it. See previous comments on boolean return variables. Notice that the `__call__` method must return a numerical value, hence my previous suggestion on doing:

```python
def __call__(...):
    return 0 if in_penumbra is True else 1
```

Reply:
Thanks, returning a boolean value does look better...I made those changes.